src/gen_rules.ml

```diff
@@ -376,6 +384,8 @@ module Gen(P : Params) = struct
         let src = lib_archive lib ~dir ~ext:(Mode.compiled_lib_ext Native) in
         let dst = lib_archive lib ~dir ~ext:".cmxs" in
         let build =
+          Build.dyn_paths (Build.arr (fun () -> [lib_archive lib ~dir ~ext:ctx.ext_lib]))
```

Comment:
My understanding of jbuilder's internals here is a bit shaky, so I'm curious:

why do you need `Build.dyn_paths` over `Build.paths`?

Reply:
I *think* (my understanding's often shaky too - there's a virtuous project in the making to write a big document on Jbuilder's internals, just for the exercise of thoroughly understanding it all oneself!) that it's because it's a generated target - so `paths` is things which should exist already (sources, etc.) and `dyn_paths` is things which are generated.

However, my change is based more on the fact that the .cmx and .cmxa files elsewhere are incorporated using dyn_paths (via Arg_spec, which generates that part of the build arrow as a side effect).

Reply:
Rules are build arrow of types `(unit, Action.t) Build.t`, but really they could be `(unit, { runtime_deps : Path.Set.t; action : Action.t }) Build.t`. It is however simpler to build the set of runtime dependencies implicitly since all we do with it is always pass it through or extend it.

In the expanded version, `Build.dyn_paths l` basically means: `Build.arr (fun x -> { x with runtime_deps = Path.Set.union x.runtime_deps (Path.Set.of_list l) })`. In any case, `Build.paths l` is always the same as `Build.dyn_paths (Build.arr (fun _ -> l))`.

The only difference between the two is that with `Build.paths`, the dependencies are known statically. This means that we can get most of the DAG without running any command and this is enough to generate the list of package dependencies that goes into the opam file. We use that feature to generate the opam files of all Jane Street packages without hassle.

Additionally, in some cases using `Build.paths` can improve compilation times as static dependencies are known earlier.

In the end, we should use `Build.paths` wherever possible.